qutip/models/piqs.py

```diff
@@ -0,0 +1,1201 @@
+#    HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+#    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+#    LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+#    DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+#    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#    (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+#    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+###############################################################################
+"""Permutational Invariant Quantum Solver (PIQS)
+
+This module calculates the Liouvillian for the dynamics of ensembles of
+identical two-level systems (TLS) in the presence of local and collective
+processes by exploiting permutational symmetry and using the Dicke basis.
+"""
+
+# Authors: Nathan Shammah, Shahnawaz Ahmed
+# Contact: shahnawaz.ahmed95@gmail.com
+
+from math import factorial
+from decimal import Decimal
+
+import numpy as np
+
+from scipy import constants
+from scipy.sparse import dok_matrix, block_diag
+
+from qutip import Qobj, spre, spost, tensor, identity, ket2dm
+from qutip import sigmax, sigmay, sigmaz, sigmap, sigmam
+from qutip.cy.piqs import Dicke as _Dicke
+from qutip.cy.piqs import (jmm1_dictionary, _num_dicke_states,
+                           _num_dicke_ladders, get_blocks)
+
+
+# Functions necessary to generate the Lindbladian/Liouvillian
+def num_dicke_states(N):
+    """Calculate the number of Dicke states.
+
+    Parameters
+    ----------
+    N: int
+        The number of two-level systems.
+
+    Returns
+    -------
+    nds: int
+        The number of Dicke states.
+    """
+    return _num_dicke_states(N)
+
+
+def num_dicke_ladders(N):
+    """Calculate the total number of ladders in the Dicke space.
+
+    For a collection of N two-level systems it counts how many different
+    "j" exist or the number of blocks in the block-diagonal matrix.
+
+    Parameters
+    ----------
+    N: int
+        The number of two-level systems.
+
+    Returns
+    -------
+    Nj: int
+        The number of Dicke ladders.
+    """
+    return _num_dicke_ladders(N)
+
+
+def num_tls(nds):
+    """Calculate the number of two-level systems.
+
+    Parameters
+    ----------
+    nds: int
+         The number of Dicke states.
+
+    Returns
+    -------
+    N: int
+        The number of two-level systems.
+    """
+    if np.sqrt(nds).is_integer():
+        # N is even
+        N = 2*(np.sqrt(nds)-1)
+    else:
+        # N is odd
+        N = 2*(np.sqrt(nds + 1/4)-1)
+    return int(N)
+
+
+class Dicke(object):
+    """The Dicke class which builds the Lindbladian and Liouvillian matrix.
+
+    Example
+    -------
+    >> from qutip.models.piqs import Dicke, j_algebra
+    >> N = 2
+    >> jx, jy, jz, jp, jm = j_algebra(N)
+    >> ensemble = Dicke(N, emission=1.)
+    >> L = ensemble.liouvillian()
+
+    Parameters
```

Comment:
As with the cy Dicke class, these are attributes. Some are also parameters. Should have separate sections